Using `optional` always results in an explicit `undefined`

I tend to write "identity decoders" that preserve the structure of the decoded value, in order to define a type and an accompanying "type guard on steroids"[^1] in one go:

[^1]: Unlike a plain old type guard, a decoder tends to stay in sync with the type and also explains _why_ a given value is not a member.

```typescript
const decoder = …;
type T = ReturnType<typeof decoder>;
```

However, I was surprised by this:

```typescript
const decoder = fieldsAuto({
  foo: number,
  bar: optional(boolean),
});
const theValue = { foo: 5 };
expect(decoder(theValue)).to.deep.equal(theValue);
```

Mocha output:

```diff
AssertionError: expected { foo: 5, bar: undefined } to deeply equal { foo: 5 }
+ expected
- actual

 {
-  "bar": [undefined]
   "foo": 5
 }
```

Is it supposed to work this way, and/or does it have to?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,9 @@
 Note: I’m currently working on several breaking changes to tiny-decoders, but I’m trying out releasing them piece by piece. The idea is that you can either upgrade version by version only having to deal with one or a few breaking changes at a time, or wait and do a bunch of them at the same time.
 
+### Version 14.0.0 (unreleased)
+
+This release removes the `fields` function, which was deprecated in version 11.0.0. See the release notes for version 11.0.0 for how to replace `fields` with `fieldsAuto`, `chain` and custom decoders.
+
 ### Version 13.0.0 (2023-10-22)
 
 > **Warning**  

--- a/README.md
+++ b/README.md
@@ -171,13 +171,6 @@ Here’s a summary of all decoders (with slightly simplified type annotations):
 <td><code>Record&lt;string, T&gt;</code></td>
 </tr>
 <tr>
-<th><a href="#fields">fields</a></th>
-<td><pre>(callback: Function) =&gt;
-  Decoder&lt;T&gt;</pre></td>
-<td>object</td>
-<td><code>T</code></td>
-</tr>
-<tr>
 <th><a href="#fieldsauto">fieldsAuto</a></th>
 <td><pre>(mapping: {
   field1: Decoder&lt;T1&gt;,
@@ -373,72 +366,6 @@ Decodes a JSON object into a TypeScript `Record`. (Yes, this function is named a
 The passed `decoder` is for each value of the object.
 
 For example, `record(number)` decodes an object where the keys can be anything and the values are numbers (into `Record<string, number>`).
-
-### fields
-
-> **Warning**  
-> This function is deprecated. Use [fieldsAuto](#fieldsAuto) instead.
-
-```ts
-function fields<T>(
-  callback: (
-    field: <U>(key: string, decoder: Decoder<U>) => U,
-    object: Record<string, unknown>,
-  ) => T,
-  {
-    exact = "allow extra",
-    allow = "object",
-  }: {
-    exact?: "allow extra" | "throw";
-    allow?: "array" | "object";
-  } = {},
-): Decoder<T>;
-```
-
-Decodes a JSON object (or array) into any TypeScript you’d like (`T`), usually an object/interface with known fields.
-
-The type annotation is a bit overwhelming, but using `fields` isn’t super complicated. In a callback, you get a `field` function that you use to pluck out stuff from the JSON object. For example:
-
-```ts
-type User = {
-  age: number;
-  active: boolean;
-  name: string;
-  description?: string | undefined;
-  version: 1;
-};
-
-const userDecoder = fields(
-  (field): User => ({
-    // Simple field:
-    age: field("age", number),
-    // Renaming a field:
-    active: field("is_active", boolean),
-    // Combining two fields:
-    name: `${field("first_name", string)} ${field("last_name", string)}`,
-    // Optional field:
-    description: field("description", undefinedOr(string)),
-    // Hardcoded field:
-    version: 1,
-  }),
-);
-
-// Plucking a single field out of an object:
-const ageDecoder: Decoder<number> = fields((field) => field("age", number));
-```
-
-`field("key", decoder)` essentially runs `decoder(obj["key"])` but with better error messages. The nice thing about `field` is that it does _not_ return a new decoder – but the value of that field! This means that you can do for instance `const type: string = field("type", string)` and then use `type` however you want inside your callback.
-
-`object` is passed in case you need to check stuff like `"my-key" in object`.
-
-Also note that you can return any type from the callback, not just objects. If you’d rather have a tuple you could return that – see the [tuples example](examples/tuples.test.ts).
-
-The `exact` option let’s you choose between ignoring extraneous data and making it a hard error.
-
-- `"allow extra"` (default) allows extra properties on the object (or extra indexes on an array).
-- `"throw"` throws a `DecoderError` for extra properties.
-
-The `allow` option defaults to only allowing JSON objects. Set it to `"array"` if you are decoding an array.
 
 ### fieldsAuto
 
@@ -779,7 +706,7 @@ Returns a new decoder that also accepts `undefined`. Alternatively, supply a `de
 
 Notes:
 
-- Using `undefinedOr` does _not_ make a field in an object optional (except in the deprecated [fields](#fields) function). It only allows the field to be `undefined`. Similarly, using the [field](#field) function to mark a field as optional does not allow setting the field to `undefined`, only omitting it.
+- Using `undefinedOr` does _not_ make a field in an object optional. It only allows the field to be `undefined`. Similarly, using the [field](#field) function to mark a field as optional does not allow setting the field to `undefined`, only omitting it.
 - JSON does not have `undefined` (only `null`). So `undefinedOr` is more useful when you are decoding something that does not come from JSON. However, even when working with JSON `undefinedOr` still has a use: If you infer types from decoders, using `undefinedOr` on object fields results in `| undefined` for the type of the field, which allows you to assign `undefined` to it which is occasionally useful.
 
 ### nullable
@@ -1082,6 +1009,6 @@ export function either<T, U>(
 This decoder would try `decoder1` first. If it fails, go on and try `decoder2`. If that fails, present both errors. I consider this a blunt tool.
 
 - If you want either a string or a number, use [multi](#multi). This let’s you switch between any JSON types.
-- For objects that can be decoded in different ways, use [fieldsUnion](#fieldsunion). If that’s not possible, use [fields](#fields) and look for the field(s) that tell which variant you’ve got. Then run the appropriate decoder for the rest of the object.
+- For objects that can be decoded in different ways, use [fieldsUnion](#fieldsunion). If that’s not possible, see the [untagged union example](examples/untagged-union.test.ts) for how you can approach the problem.
 
 The above approaches result in a much simpler [DecoderError](#decodererror) type, and also results in much better error messages, since there’s never a need to present something like “decoding failed in the following 2 ways: …”

--- a/examples/readme.test.ts
+++ b/examples/readme.test.ts
@@ -1,4 +1,3 @@
-/* eslint-disable @typescript-eslint/no-shadow */ // TODO: Remove this line when removing the `fields` function.
 import { expectType, TypeEqual } from "ts-expect";
 import { expect, test } from "vitest";
 
@@ -8,7 +7,6 @@ import {
   Decoder,
   DecoderError,
   field,
-  fields,
   fieldsAuto,
   number,
   repr,
@@ -136,51 +134,6 @@ test("default vs sensitive error messages", () => {
     Got: number
     (Actual values are hidden in sensitive mode.)
   `);
-});
-
-test("fields", () => {
-  type User = {
-    age: number;
-    active: boolean;
-    name: string;
-    description?: string | undefined;
-    version: 1;
-  };
-
-  const userDecoder = fields(
-    (field): User => ({
-      // Simple field:
-      age: field("age", number),
-      // Renaming a field:
-      active: field("is_active", boolean),
-      // Combining two fields:
-      name: `${field("first_name", string)} ${field("last_name", string)}`,
-      // Optional field:
-      description: field("description", undefinedOr(string)),
-      // Hardcoded field:
-      version: 1,
-    }),
-  );
-
-  expect(
-    userDecoder({
-      age: 30,
-      is_active: true,
-      first_name: "John",
-      last_name: "Doe",
-    }),
-  ).toStrictEqual({
-    active: true,
-    age: 30,
-    description: undefined,
-    name: "John Doe",
-    version: 1,
-  });
-
-  // Plucking a single field out of an object:
-  const ageDecoder: Decoder<number> = fields((field) => field("age", number));
-
-  expect(ageDecoder({ age: 30 })).toBe(30);
 });
 
 test("fieldsAuto", () => {

--- a/examples/type-annotations.test.ts
+++ b/examples/type-annotations.test.ts
@@ -3,7 +3,7 @@
 
 import { expect, test } from "vitest";
 
-import { Decoder, fields, fieldsAuto, number, string } from "../";
+import { Decoder, fieldsAuto, number, string } from "../";
 
 test("type annotations", () => {
   // First, a small test type and a function that receives it:
@@ -20,43 +20,29 @@ test("type annotations", () => {
    * MISSPELLED PROPERTY
    */
 
-  // Here are two decoders for `Person`, but without explicit type annotations.
+  // Here’s a decoder for `Person`, but without an explicit type annotation.
   // TypeScript will infer what they decode into (try hovering `personDecoder1`
-  // and `personDecoder1Auto` in your editor!), but it won’t know that you
-  // intended to decode a `Person`. As you can see, I’ve misspelled `age` as `aye`.
-  const personDecoder1 = fields((field) => ({
-    name: field("name", string),
-    aye: field("age", number),
-  }));
-  const personDecoder1Auto = fieldsAuto({
+  // in your editor!), but it won’t know that you intended to decode a `Person`.
+  // As you can see, I’ve misspelled `age` as `aye`.
+  const personDecoder1 = fieldsAuto({
     name: string,
     aye: number,
   });
-  // Since TypeScript has inferred legit decoders above, it marks the following
-  // two calls as errors (you can’t pass an object with `aye` as a `Person`),
-  // while the _real_ errors of course are in the decoders themselves.
+  // Since TypeScript has inferred a legit decoder above, it marks the following
+  // call as an error (you can’t pass an object with `aye` as a `Person`),
+  // while the _real_ error of course is in the decoder itself.
   // @ts-expect-error Property 'age' is missing in type '{ name: string; aye: number; }' but required in type 'Person'.
   greet(personDecoder1(testPerson));
-  // @ts-expect-error Property 'age' is missing in type '{ name: string; aye: number; }' but required in type 'Person'.
-  greet(personDecoder1Auto(testPerson));
 
-  // The way to make the above type errors more clear is to provide explicit type
-  // annotations, so that TypeScript knows what you’re trying to do.
-  const personDecoder2 = fields(
-    (field): Person => ({
-      name: field("name", string),
-      // @ts-expect-error Object literal may only specify known properties, and 'aye' does not exist in type 'Person'.
-      aye: field("age", number),
-    }),
-  );
+  // The way to make the above type error more clear is to provide an explicit type
+  // annotation, so that TypeScript knows what you’re trying to do.
   // @ts-expect-error Type 'Decoder<{ name: string; aye: number; }, unknown>' is not assignable to type 'Decoder<Person>'.
   //   Property 'age' is missing in type '{ name: string; aye: number; }' but required in type 'Person'.ts(2322)
-  const personDecoder2Auto: Decoder<Person> = fieldsAuto({
+  const personDecoder2: Decoder<Person> = fieldsAuto({
     name: string,
     aye: number,
   });
   greet(personDecoder2(testPerson));
-  greet(personDecoder2Auto(testPerson));
 
   /*
    * EXTRA PROPERTY
@@ -64,76 +50,42 @@ test("type annotations", () => {
 
   // TypeScript allows passing extra properties, so without type annotations
   // there are no errors:
-  const personDecoder5 = fields((field) => ({
-    name: field("name", string),
-    age: field("age", number),
-    extra: field("extra", string),
-  }));
-  const personDecoder5Auto = fieldsAuto({
+  const personDecoder3 = fieldsAuto({
     name: string,
     age: number,
     extra: string,
   });
-  // These would ideally complain about the extra property, but they don’t.
-  greet(personDecoder5(testPerson));
-  greet(personDecoder5Auto(testPerson));
+  // This would ideally complain about the extra property, but it doesn’t.
+  greet(personDecoder3(testPerson));
 
   // Adding `Decoder<Person>` does not seem to help TypeScript find any errors:
-  const personDecoder6: Decoder<Person> = fields((field) => ({
-    name: field("name", string),
-    age: field("age", number),
-    extra: field("extra", string),
-  }));
-  const personDecoder6Auto: Decoder<Person> = fieldsAuto({
+  const personDecoder4: Decoder<Person> = fieldsAuto({
     name: string,
     age: number,
     extra: string,
   });
-  greet(personDecoder6(testPerson));
-  greet(personDecoder6Auto(testPerson));
+  greet(personDecoder4(testPerson));
 
-  // The recommended type annotation for `fields` does produce errors!
-  const personDecoder7 = fields(
-    (field): Person => ({
-      name: field("name", string),
-      age: field("age", number),
-      // @ts-expect-error Object literal may only specify known properties, and 'extra' does not exist in type 'Person'.
-      extra: field("extra", string),
-    }),
-  );
-  const personDecoder7Auto: Decoder<Person> = fieldsAuto({
+  // This is currently not an error unfortunately, but in a future version of tiny-decoders it will be.
+  const personDecoder5: Decoder<Person> = fieldsAuto({
     name: string,
     age: number,
-    // This is currently not an error unfortunately, but in a future version of tiny-decoders it will be.
+    // Here is where the error will be.
     extra: string,
   });
-  greet(personDecoder7(testPerson));
-  greet(personDecoder7Auto(testPerson));
+  greet(personDecoder5(testPerson));
   // See these TypeScript issues for more information:
   // https://github.com/microsoft/TypeScript/issues/7547
   // https://github.com/microsoft/TypeScript/issues/18020
 
-  // Finally, some compiling decoders.
-  const personDecoder8 = fields(
-    (field): Person => ({
-      name: field("name", string),
-      age: field("age", number),
-    }),
-  );
-  const personDecoder8Auto: Decoder<Person> = fieldsAuto({
+  // Finally, a compiling decoder.
+  const personDecoder6: Decoder<Person> = fieldsAuto({
     name: string,
     age: number,
   });
-  greet(personDecoder8(testPerson));
-  greet(personDecoder8Auto(testPerson));
+  greet(personDecoder6(testPerson));
 
-  expect(personDecoder8(testPerson)).toMatchInlineSnapshot(`
-    {
-      "age": 30,
-      "name": "John",
-    }
-  `);
-  expect(personDecoder8Auto(testPerson)).toMatchInlineSnapshot(`
+  expect(personDecoder6(testPerson)).toMatchInlineSnapshot(`
     {
       "age": 30,
       "name": "John",

--- a/examples/type-inference.test.ts
+++ b/examples/type-inference.test.ts
@@ -1,4 +1,3 @@
-/* eslint-disable @typescript-eslint/no-shadow */ // TODO: Remove this line when removing the `fields` function.
 // This file shows how to infer types from decoders.
 
 import { expectType, TypeEqual } from "ts-expect";
@@ -7,17 +6,12 @@ import { expect, test } from "vitest";
 import {
   boolean,
   chain,
-  DecoderError,
   field,
-  fields,
   fieldsAuto,
-  fieldsUnion,
   multi,
   number,
   string,
   stringUnion,
-  tag,
-  undefinedOr,
 } from "..";
 
 test("making a type from a decoder", () => {
@@ -25,29 +19,14 @@ test("making a type from a decoder", () => {
   // typing the same thing again in the decoder (especially `fieldsAuto` decoders
   // look almost identical to `type` they decode to!), you can start with the
   // decoder and extract the type afterwards with TypeScript’s `ReturnType` utility.
-  const personDecoder1 = fields((field) => ({
-    name: field("name", string),
-    age: field("age", number),
-  }));
-  const personDecoder1Auto = fieldsAuto({
+  const personDecoder = fieldsAuto({
     name: string,
     age: number,
   });
 
-  // Hover over `Person1` to see what it looks like!
-  type Person1 = ReturnType<typeof personDecoder1>;
-  expectType<TypeEqual<Person1, { name: string; age: number }>>(true);
-
-  // Hover over `Person1Auto` to see what it looks like!
-  type Person1Auto = ReturnType<typeof personDecoder1Auto>;
-  expectType<TypeEqual<Person1Auto, { name: string; age: number }>>(true);
-
-  // A little sanity check that it actually works:
-  const testPerson: unknown = { name: "John", age: 30 };
-  function meet(personA: Person1, personB: Person1Auto): string {
-    return `${personA.name} meets ${personB.name}`;
-  }
-  meet(personDecoder1(testPerson), personDecoder1Auto(testPerson));
+  // Hover over `Person` to see what it looks like!
+  type Person = ReturnType<typeof personDecoder>;
+  expectType<TypeEqual<Person, { name: string; age: number }>>(true);
 
   // If it feels like you are specifying everything twice – once in a `type` or
   // `interface`, and once in the decoder – you might find this `ReturnType`
@@ -112,275 +91,6 @@ test("making a type from a decoder", () => {
     extra: "prop",
   };
   expect(user3).toMatchObject(user);
-
-  // Here’s the same decoder again, but written using `fields` instead of
-  // `fieldsAuto`. It should give the same inferred type.
-  const userDecoder2 = fields((field) => ({
-    id: field(
-      "id",
-      chain(multi(["string", "number"]), ({ value }) => value),
-    ),
-    name: field("name", string),
-    age: field("age", number),
-    active: field("active", boolean),
-    country: field("country", undefinedOr(string)),
-    type: field("type", stringUnion(["user"])),
-  }));
-
-  type User2 = ReturnType<typeof userDecoder2>;
-
-  expectType<TypeEqual<User, User2>>(true);
-
-  const user4: User2 = {
-    id: 1,
-    name: "John Doe",
-    age: 30,
-    active: true,
-    country: undefined,
-    // @ts-expect-error Type '"nope"' is not assignable to type '"user"'.
-    type: "nope",
-  };
-  expect({ ...user4, type: "user" }).toMatchObject(user);
-});
-
-test("making a type from a decoder – unions", () => {
-  // Let’s say we need to support two types of users – anonymous and registered
-  // ones. This is where `fieldsUnion` shines! It’s both easier to use and gives
-  // a better inferred type!
-  const userDecoder1 = fieldsUnion("type", [
-    {
-      type: tag("anonymous"),
-      sessionId: number,
-    },
-    {
-      type: tag("registered"),
-      id: number,
-      name: string,
-    },
-  ]);
-  type InferredType1 = ReturnType<typeof userDecoder1>;
-  type ExpectedType1 =
-    | { type: "anonymous"; sessionId: number }
-    | { type: "registered"; id: number; name: string };
-  expectType<TypeEqual<InferredType1, ExpectedType1>>(true);
-
-  // For comparison, let’s do the same decoder “manually”.
-  // This might not be super interesting – but I’ve kept this “TypeScript
-  // research” for my future self.
-  // Unfortunately, TypeScript doesn’t infer the type you might have expected:
-  const userDecoder2 = fields((field) => {
-    const type = field("type", string);
-
-    switch (type) {
-      case "anonymous":
-        return {
-          type: "anonymous",
-          sessionId: field("sessionId", number),
-        };
-
-      case "registered":
-        return {
-          type: "registered",
-          id: field("id", number),
-          name: field("name", string),
-        };
-
-      default:
-        throw new DecoderError({
-          message: "Unknown user type",
-          value: type,
-          key: "type",
-        });
-    }
-  });
-  type InferredType2 = ReturnType<typeof userDecoder2>;
-  type ActualType2 =
-    | {
-        type: string;
-        id: number;
-        name: string;
-        sessionId?: undefined;
-      }
-    | {
-        type: string;
-        sessionId: number;
-        id?: undefined;
-        name?: undefined;
-      };
-  expectType<TypeEqual<InferredType2, ActualType2>>(true);
-
-  // To turn `type: string` into `type: "anonymous"` and `type: "registered"`, add
-  // `as const`:
-  const userDecoder3 = fields((field) => {
-    const type = field("type", string);
-
-    switch (type) {
-      case "anonymous":
-        return {
-          type: "anonymous" as const,
-          sessionId: field("sessionId", number),
-        };
-
-      case "registered":
-        return {
-          type: "registered" as const,
-          id: field("id", number),
-          name: field("name", string),
-        };
-
-      default:
-        throw new DecoderError({
-          message: "Unknown user type",
-          value: type,
-          key: "type",
-        });
-    }
-  });
-  type InferredType3 = ReturnType<typeof userDecoder3>;
-  type ActualType3 =
-    | {
-        type: "anonymous";
-        sessionId: number;
-        id?: undefined;
-        name?: undefined;
-      }
-    | {
-        type: "registered";
-        id: number;
-        name: string;
-        sessionId?: undefined;
-      };
-  expectType<TypeEqual<InferredType3, ActualType3>>(true);
-
-  // However, `id?: undefined` and similar are still part of the type. They don’t
-  // hurt super much when it comes to type safety. They are optional fields that
-  // only are allowed to be set to `undefined` – you can’t do much with that.
-  // However, they do clutter tooltips and autocomplete in your editor.
-  const testUser = userDecoder3({ type: "anonymous", sessionId: 123 });
-  if (testUser.type === "anonymous") {
-    // Type `testUser.` above this line and check your editor’s autocomplete.
-    // You probably get something like this:
-    //
-    // - id?: undefined
-    // - name?: undefined
-    // - sessionId: number
-    // - type: "anonymous"
-    //
-    // And if you hover over `testUser` in the line you typed, the tooltip showing
-    // the type might say:
-    //
-    //     {
-    //       type: "anonymous";
-    //       sessionId: number;
-    //       id?: undefined;
-    //       name?: undefined;
-    //     }
-    //
-    // Those extra fields are just noisy. They exist because of this:
-    // https://github.com/microsoft/TypeScript/pull/19513
-    //
-    // But can we get rid of them?
-  }
-
-  // Turns out we can get rid of the extra properties by using the good old
-  // `identity` function! It’s just a function that returns whatever is passed to
-  // it. A bit of an ugly workaround, but it works.
-  const id = <T>(x: T): T => x;
-
-  // And when wrapping each `return` in `id(...)` the extra properties vanish!
-  const userDecoder4 = fields((field) => {
-    const type = field("type", string);
-
-    switch (type) {
-      case "anonymous":
-        return id({
-          type: "anonymous" as const,
-          sessionId: field("sessionId", number),
-        });
-
-      case "registered":
-        return id({
-          type: "registered" as const,
-          id: field("id", number),
-          name: field("name", string),
-        });
-
-      default:
-        throw new DecoderError({
-          message: "Unknown user type",
-          value: type,
-          key: "type",
-        });
-    }
-  });
-  type InferredType4 = ReturnType<typeof userDecoder4>;
-  type ActualType4 =
-    | {
-        type: "anonymous";
-        sessionId: number;
-      }
-    | {
-        type: "registered";
-        id: number;
-        name: string;
-      };
-  expectType<TypeEqual<InferredType4, ActualType4>>(true);
-
-  // Another way is to write separate decoders for each case.
-  // Note how I disabled the explicit-function-return-type rule – inferring the
-  // type is the whole point, and having to disable ESLint rules is annoying…
-  // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
-  function getUserDecoder(type: unknown) {
-    switch (type) {
-      case "anonymous":
-        return fields((field) => ({
-          type: "anonymous" as const,
-          sessionId: field("sessionId", number),
-        }));
-
-      case "registered":
-        // You can also use `fieldsAuto`:
-        return fieldsAuto({
-          type: () => "registered" as const,
-          id: number,
-          name: string,
-        });
-
-      default:
-        throw new DecoderError({
-          message: "Unknown user type",
-          value: type,
-          key: "type",
-        });
-    }
-  }
-
-  // This “decodes” the "type" field using `getUserDecoder`. Remember that a
-  // decoder is allowed to return whatever it wants. `getUserDecoder` returns a
-  // _new_ decoder, which we immediately call.
-  const userDecoder5 = fields((field, obj) =>
-    field("type", getUserDecoder)(obj),
-  );
-  type InferredType5 = ReturnType<typeof userDecoder5>;
-  type ActualType5 =
-    | {
-        type: "anonymous";
-        sessionId: number;
-      }
-    | {
-        type: "registered";
-        id: number;
-        name: string;
-      };
-  expectType<TypeEqual<InferredType5, ActualType5>>(true);
-
-  expect(userDecoder5({ type: "anonymous", sessionId: 1 }))
-    .toMatchInlineSnapshot(`
-      {
-        "sessionId": 1,
-        "type": "anonymous",
-      }
-    `);
 });
 
 test("making a type from an object and stringUnion", () => {

--- a/examples/untagged-union.test.ts
+++ b/examples/untagged-union.test.ts
@@ -1,0 +1,124 @@
+import { expect, test } from "vitest";
+
+import {
+  array,
+  boolean,
+  chain,
+  Decoder,
+  fieldsAuto,
+  number,
+  string,
+  undefinedOr,
+} from "../";
+
+test("untagged union", () => {
+  type User = {
+    name: string;
+    followers: number;
+  };
+
+  type Failure = {
+    error: string;
+    errorCode: number;
+  };
+
+  type UserResult = Failure | User;
+
+  const userDecoder = fieldsAuto({
+    name: string,
+    followers: number,
+  });
+
+  const failureDecoder = fieldsAuto({
+    error: string,
+    errorCode: number,
+  });
+
+  const userResultDecoder: Decoder<UserResult> = (value) =>
+    // This is a bit annoying to do. Prefer a tagged union and use `fieldsAuto`.
+    // But when that’s not possible, this is a simple way of “committing” to one
+    // of the union variants and choosing a decoder based on that.
+    // This approach results in much easier to understand error messages at
+    // runtime than an approach of first trying the first decoder, and then
+    // the second (because if both fail, you need to display both error messages).
+    typeof value === "object" && value !== null && "error" in value
+      ? failureDecoder(value)
+      : userDecoder(value);
+
+  expect(userResultDecoder({ name: "John", followers: 42 }))
+    .toMatchInlineSnapshot(`
+    {
+      "followers": 42,
+      "name": "John",
+    }
+  `);
+
+  expect(userResultDecoder({ error: "Not found", errorCode: 404 }))
+    .toMatchInlineSnapshot(`
+    {
+      "error": "Not found",
+      "errorCode": 404,
+    }
+  `);
+});
+
+test("tagged union, but using boolean instead of string", () => {
+  const adminDecoder = chain(
+    fieldsAuto({
+      name: string,
+      access: array(string),
+    }),
+    (props) => ({
+      isAdmin: true as const,
+      ...props,
+    }),
+  );
+
+  const notAdminDecoder = chain(
+    fieldsAuto({
+      name: string,
+      location: undefinedOr(string),
+    }),
+    (props) => ({
+      isAdmin: false as const,
+      ...props,
+    }),
+  );
+
+  type User =
+    | ReturnType<typeof adminDecoder>
+    | ReturnType<typeof notAdminDecoder>;
+
+  const userDecoder: Decoder<User> = (value) => {
+    const { isAdmin } = fieldsAuto({ isAdmin: boolean })(value);
+    return isAdmin ? adminDecoder(value) : notAdminDecoder(value);
+  };
+
+  expect(
+    userDecoder({
+      isAdmin: true,
+      name: "John",
+      access: [],
+    }),
+  ).toMatchInlineSnapshot(`
+    {
+      "access": [],
+      "isAdmin": true,
+      "name": "John",
+    }
+  `);
+
+  expect(
+    userDecoder({
+      isAdmin: false,
+      name: "Jane",
+      location: undefined,
+    }),
+  ).toMatchInlineSnapshot(`
+    {
+      "isAdmin": false,
+      "location": undefined,
+      "name": "Jane",
+    }
+  `);
+});

--- a/index.ts
+++ b/index.ts
@@ -4,18 +4,6 @@
 
 export type Decoder<T, U = unknown> = (value: U) => T;
 
-type WithUndefinedAsOptional<T> = T extends Record<string, unknown>
-  ? Expand<{ [P in OptionalKeys<T>]?: T[P] } & { [P in RequiredKeys<T>]: T[P] }>
-  : T;
-
-type RequiredKeys<T> = {
-  [P in keyof T]: undefined extends T[P] ? never : P;
-}[keyof T];
-
-type OptionalKeys<T> = {
-  [P in keyof T]: undefined extends T[P] ? P : never;
-}[keyof T];
-
 // Make VSCode show `{ a: string; b?: number }` instead of `{ a: string } & { b?: number }`.
 // https://stackoverflow.com/a/57683652/2010616
 type Expand<T> = T extends infer O ? { [K in keyof O]: O[K] } : never;
@@ -104,58 +92,6 @@ export function record<T>(decoder: Decoder<T>): Decoder<Record<string, T>> {
     }
 
     return result;
-  };
-}
-
-/**
- * @deprecated Use `fieldsAuto` instead.
- */
-export function fields<T>(
-  callback: (
-    field: <U>(key: string, decoder: Decoder<U>) => U,
-    object: Record<string, unknown>,
-  ) => T,
-  {
-    exact = "allow extra",
-    allow = "object",
-  }: {
-    exact?: "allow extra" | "throw";
-    allow?: "array" | "object";
-  } = {},
-): Decoder<WithUndefinedAsOptional<T>> {
-  return function fieldsDecoder(value: unknown): WithUndefinedAsOptional<T> {
-    const object: Record<string, unknown> =
-      allow === "array"
-        ? (unknownArray(value) as unknown as Record<string, unknown>)
-        : unknownRecord(value);
-    const knownFields = Object.create(null) as Record<string, null>;
-
-    function field_<U>(key: string, decoder: Decoder<U>): U {
-      try {
-        const result = decoder(object[key]);
-        knownFields[key] = null;
-        return result;
-      } catch (error) {
-        throw DecoderError.at(error, key);
-      }
-    }
-
-    const result = callback(field_, object);
-
-    if (exact !== "allow extra") {
-      const unknownFields = Object.keys(object).filter(
-        (key) => !Object.prototype.hasOwnProperty.call(knownFields, key),
-      );
-      if (unknownFields.length > 0) {
-        throw new DecoderError({
-          tag: "exact fields",
-          knownFields: Object.keys(knownFields),
-          got: unknownFields,
-        });
-      }
-    }
-
-    return result as WithUndefinedAsOptional<T>;
   };
 }
 

--- a/tests/decoders.test.ts
+++ b/tests/decoders.test.ts
@@ -1,5 +1,3 @@
-/* eslint-disable @typescript-eslint/no-shadow */ // TODO: Remove this line when removing the `fields` function.
-
 import { expectType, TypeEqual } from "ts-expect";
 import { describe, expect, test } from "vitest";
 
@@ -10,7 +8,6 @@ import {
   Decoder,
   DecoderError,
   field,
-  fields,
   fieldsAuto,
   fieldsUnion,
   multi,
@@ -262,265 +259,6 @@ describe("record", () => {
     expect(
       run(record(number), JSON.parse(`{"a": 1, "__proto__": 2, "b": 3}`)),
     ).toStrictEqual({ a: 1, b: 3 });
-  });
-});
-
-describe("fields", () => {
-  test("basic", () => {
-    type Person = ReturnType<typeof personDecoder>;
-    const personDecoder = fields((field) => ({
-      id: field("id", number),
-      firstName: field("first_name", string),
-    }));
-
-    expectType<TypeEqual<Person, { id: number; firstName: string }>>(true);
-    expectType<Person>(personDecoder({ id: 1, first_name: "John" }));
-
-    expect(personDecoder({ id: 1, first_name: "John" })).toStrictEqual({
-      id: 1,
-      firstName: "John",
-    });
-
-    expect(run(personDecoder, { id: "1", first_name: "John" }))
-      .toMatchInlineSnapshot(`
-        At root["id"]:
-        Expected a number
-        Got: "1"
-      `);
-
-    expect(run(personDecoder, { id: 1, firstName: "John" }))
-      .toMatchInlineSnapshot(`
-        At root["first_name"]:
-        Expected a string
-        Got: undefined
-      `);
-
-    expect(
-      run(
-        fields((field) => field("0", number)),
-        [1],
-      ),
-    ).toMatchInlineSnapshot(`
-      At root:
-      Expected an object
-      Got: [
-        1
-      ]
-    `);
-  });
-
-  describe("exact", () => {
-    test("allows excess properties by default", () => {
-      expect(
-        run(
-          fields((field) => [field("one", string), field("two", boolean)]),
-          { one: "a", two: true, three: 3, four: {} },
-        ),
-      ).toStrictEqual(["a", true]);
-      expect(
-        run(
-          fields((field) => [field("one", string), field("two", boolean)], {
-            exact: "allow extra",
-          }),
-          { one: "a", two: true, three: 3, four: {} },
-        ),
-      ).toStrictEqual(["a", true]);
-    });
-
-    test("throw on excess properties", () => {
-      expect(
-        run(
-          fields((field) => [field("one", string), field("two", boolean)], {
-            exact: "throw",
-          }),
-          { one: "a", two: true, three: 3, four: {} },
-        ),
-      ).toMatchInlineSnapshot(`
-        At root:
-        Expected only these fields:
-          "one",
-          "two"
-        Found extra fields:
-          "three",
-          "four"
-      `);
-    });
-
-    test("large number of excess properties", () => {
-      expect(
-        run(
-          fields((field) => [field("1", boolean), field("2", boolean)], {
-            exact: "throw",
-          }),
-          Object.fromEntries(Array.from({ length: 100 }, (_, i) => [i, false])),
-        ),
-      ).toMatchInlineSnapshot(`
-        At root:
-        Expected only these fields:
-          "1",
-          "2"
-        Found extra fields:
-          "0",
-          "3",
-          "4",
-          "5",
-          "6",
-          (93 more)
-      `);
-    });
-
-    test("different fields based on one field", () => {
-      const userDecoder = fields(
-        (field) =>
-          field("isAdmin", boolean)
-            ? {
-                isAdmin: true,
-                name: field("name", string),
-                access: field("access", array(string)),
-              }
-            : {
-                isAdmin: false,
-                name: field("name", string),
-                location: undefinedOr(string),
-              },
-        { exact: "throw" },
-      );
-
-      expect(
-        run(userDecoder, {
-          isAdmin: true,
-          name: "John",
-          access: [],
-          age: 12,
-        }),
-      ).toMatchInlineSnapshot(`
-        At root:
-        Expected only these fields:
-          "isAdmin",
-          "name",
-          "access"
-        Found extra fields:
-          "age"
-      `);
-
-      expect(
-        run(userDecoder, {
-          isAdmin: false,
-          name: "Jane",
-          access: [],
-          age: 12,
-        }),
-      ).toMatchInlineSnapshot(`
-        At root:
-        Expected only these fields:
-          "isAdmin",
-          "name"
-        Found extra fields:
-          "access",
-          "age"
-      `);
-    });
-  });
-
-  describe("allow", () => {
-    test("object", () => {
-      const decoder = fields((field) => field("length", number), {
-        allow: "object",
-      });
-      expect(decoder({ length: 0 })).toBe(0);
-      expect(run(decoder, [])).toMatchInlineSnapshot(`
-        At root:
-        Expected an object
-        Got: []
-      `);
-    });
-
-    test("array", () => {
-      const decoder = fields(
-        (field) => [field("length", number), field("0", boolean)],
-        { allow: "array" },
-      );
-      expect(decoder([true])).toStrictEqual([1, true]);
-      expect(run(decoder, { length: 0 })).toMatchInlineSnapshot(`
-        At root:
-        Expected an array
-        Got: {
-          "length": 0
-        }
-      `);
-      expect(run(decoder, [])).toMatchInlineSnapshot(`
-        At root["0"]:
-        Expected a boolean
-        Got: undefined
-      `);
-
-      // @ts-expect-error Argument of type 'number' is not assignable to parameter of type 'string'.
-      fields((field) => field(0, string));
-    });
-  });
-
-  test("obj", () => {
-    const objInput = {};
-    const result = fields((_field, obj) => {
-      expect(obj).toBe(objInput);
-      return 1;
-    })(objInput);
-    expect(result).toBe(1);
-  });
-
-  test("empty object", () => {
-    const decoder = fields(() => 1, { exact: "throw" });
-    expect(decoder({})).toBe(1);
-    expect(run(decoder, { a: 1 })).toMatchInlineSnapshot(`
-      At root:
-      Expected only these fields: (none)
-      Found extra fields:
-        "a"
-    `);
-  });
-
-  test("return a class", () => {
-    class Person {
-      constructor(
-        public firstName: string,
-        public lastName: string,
-        public age?: number,
-      ) {
-        this.firstName = firstName;
-        this.lastName = lastName;
-        this.age = age;
-      }
-
-      getFullName(): string {
-        return `${this.firstName} ${this.lastName}`;
-      }
-    }
-
-    const decoder = fields(
-      (field) =>
-        new Person(
-          field("first_name", string),
-          field("last_name", string),
-          field("age", undefinedOr(number)),
-        ),
-    );
-
-    const person = decoder({ first_name: "John", last_name: "Doe" });
-
-    if (person.firstName !== "John") {
-      // @ts-expect-error Object is possibly 'undefined'.
-      person.age.toFixed(2);
-    }
-
-    expect(person).toMatchInlineSnapshot(`
-      Person {
-        "age": undefined,
-        "firstName": "John",
-        "lastName": "Doe",
-      }
-    `);
-
-    expect(person.getFullName()).toMatchInlineSnapshot(`"John Doe"`);
   });
 });
 
@@ -1326,41 +1064,6 @@ describe("undefinedOr", () => {
     expect(decoder("a")).toBe("a");
   });
 
-  test("using with fields results in an optional field", () => {
-    type Person = ReturnType<typeof personDecoder>;
-    const personDecoder = fields((field) => ({
-      name: field("name", string),
-      age: field("age", undefinedOr(number)),
-    }));
-
-    expectType<TypeEqual<Person, { name: string; age?: number }>>(true);
-
-    expect(personDecoder({ name: "John" })).toStrictEqual({
-      name: "John",
-      age: undefined,
-    });
-
-    expect(personDecoder({ name: "John", age: undefined })).toStrictEqual({
-      name: "John",
-      age: undefined,
-    });
-
-    expect(personDecoder({ name: "John", age: 45 })).toStrictEqual({
-      name: "John",
-      age: 45,
-    });
-
-    expect(run(personDecoder, { name: "John", age: "old" }))
-      .toMatchInlineSnapshot(`
-        At root["age"] (optional):
-        Expected a number
-        Got: "old"
-      `);
-
-    const person: Person = { name: "John" };
-    void person;
-  });
-
   test("using with fieldsAuto does NOT result in an optional field", () => {
     type Person = ReturnType<typeof personDecoder>;
     const personDecoder = fieldsAuto({
@@ -1489,17 +1192,19 @@ describe("nullable", () => {
 
   test("nullable field", () => {
     type Person = ReturnType<typeof personDecoder>;
-    const personDecoder = fields((field) => ({
-      name: field("name", string),
-      age: field("age", nullable(number)),
-    }));
+    const personDecoder = fieldsAuto({
+      name: string,
+      age: nullable(number),
+    });
 
     expectType<TypeEqual<Person, { name: string; age: number | null }>>(true);
 
     expect(run(personDecoder, { name: "John" })).toMatchInlineSnapshot(`
-      At root["age"] (nullable):
-      Expected a number
-      Got: undefined
+      At root:
+      Expected an object with a field called: "age"
+      Got: {
+        "name": "John"
+      }
     `);
 
     expect(run(personDecoder, { name: "John", age: undefined }))

